boards: Enable GPIO by default on boards with nRF SoCs

anangl opened the pull request:
This patch enables the GPIO driver by default on all boards equipped
with an nRF SoC (all boards having `CONFIG_SOC_FAMILY_NRF=y` in their
`_defconfig` file).
In vast majority of cases the driver is needed, so it is more
conveninent to enable it at board level than in particular
applications
And if the driver is undesired for some reason, it can be still
disabled in the application config.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
boards: Enable GPIO by default on boards with nRF SoCs

This patch enables the GPIO driver by default on all boards equipped
with an nRF SoC (all boards having `CONFIG_SOC_FAMILY_NRF=y` in their
`_defconfig` file).
In vast majority of cases the driver is needed, so it is more
convenient to enable it at board level than in particular
applications.
And if the driver is undesired for some reason, it can be still
disabled in the application config.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented:
cc @erwango

anangl commented:
Labeled as a bugfix, since the [initialization code for nrf52840_pca10090](https://github.com/zephyrproject-rtos/zephyr/blob/master/boards/arm/nrf52840_pca10090/board.c) requires the GPIO driver, and currently it is not enabled, resulting in a hard fault if assertions are note enabled.

galak commented:
> Labeled as a bugfix, since the [initialization code for nrf52840_pca10090](https://github.com/zephyrproject-rtos/zephyr/blob/master/boards/arm/nrf52840_pca10090/board.c) requires the GPIO driver, and currently it is not enabled, resulting in a hard fault if assertions are note enabled.

If the init code for nrf52840_pca10090 requires GPIO than I think you should have a separate PR just for it as a bug fix.  I can't remember if we can 'select GPIO' in the `boards/arm/nrf52840_pca10090/Kconfig.board` as that would make the most sense.

I'll let @erwango comment on generally enabling GPIO, but I feel like we shouldn't be doing that if its not needed.

anangl commented:
> I'll let @erwango comment on generally enabling GPIO, but I feel like we shouldn't be doing that if its not needed.

After reading @erwango's https://github.com/zephyrproject-rtos/zephyr/pull/14655#issuecomment-476639114, I thought that the GPIO driver should be treated as a "basic" driver that is by default enabled for generally all boards, similarly to the SERIAL driver. Actually, for all the boards affected by this PR, it looked like this earlier - the GPIO driver was by default enabled, but it was done implicitly, by enabling the SERIAL driver that in turn `select`ed the GPIO one (see my https://github.com/zephyrproject-rtos/zephyr/pull/14655#issuecomment-476676943). The situation changed after [this commit](https://github.com/zephyrproject-rtos/zephyr/commit/7c13ba9965338d2bb8ca0d13847e110a5e7fa67d#diff-cb43ddfba4a214d29f2732e7fa1a402c) was merged. Thus, there might be more cases like the nrf52840_pca10090 one I mentioned above, that something that worked properly before no longer works because the GPIO driver is not enabled.

erwango commented:
> I'll let @erwango comment on generally enabling GPIO, but I feel like we shouldn't be doing that if its not needed.

As mentioned by @anangl I commented on other PR that GPIO should be part of basic drivers that should be enabled by default such as CLOCK_CONTROL and SERIAL. There are quite little things we can do when not activated and this may be difficult to find out for new comer. If really not desired for some reason, it up to application to disabled.

galak commented on boards/arm/nrf52840_pca10090/nrf52840_pca10090_defconfig:
```diff
@@ -7,6 +7,9 @@ CONFIG_BOARD_NRF52840_PCA10090=y
 # Enable MPU
 CONFIG_ARM_MPU=y
 
+# enable GPIO
```
Since this board requires GPIO to function, add 'select GPIO' in boards/arm/nrf52840_pca10090/Kconfig.board and remove setting it in the defconfig.

MaureenHelm replied:
Does the board actually require GPIO to function, or does the board init code just not properly handle the case that GPIO is not enabled?

anangl replied:
GPIO is required by this board. When the driver is not enabled, this board's initialization code cannot configure the on-board switches (that provide routing of certain signals) according to the configuration set in Kconfig, hence the code become useless.

galak reviewed (changes requested):
Update boards/arm/nrf52840_pca10090 to select GPIO

anangl committed:
boards: nrf52840_pca10090: Force GPIO to be always enabled

The GPIO driver is required by this board's initialization code, hence
it is forced to be enabled always, not only enabled by default like on
other boards equipped with an nRF SoC.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented:
@galak
> Update boards/arm/nrf52840_pca10090 to select GPIO

Done.

galak commented on boards/arm/nrf52840_pca10090/nrf52840_pca10090_defconfig:
```diff
@@ -7,6 +7,11 @@ CONFIG_BOARD_NRF52840_PCA10090=y
 # Enable MPU
 CONFIG_ARM_MPU=y
 
+# The GPIO driver is required by this board's initialization code (board.c),
```
Remove from defconfig, it will get set by the select, so you don't need it here.

anangl replied:
@galak I intentionally commented out the `CONFIG_GPIO=y` line instead of removing it completely, and I added the comment explaining why `GPIO` is not enabled here. I wanted to leave a trace in case someone searched for `CONFIG_GPIO=y` in boards, for instance.
You want me to remove this commented out line? Or the whole comment?

galak replied:
I think we should remove it.

anangl replied:
Removed.

lemrey reviewed (approved):
LGTM.
